Simplify tool implemetations

pakrym-oai opened the pull request:
Use Result<String, FunctionCallError> for all tool handling code and rely on error propagation instead of creating failed items everywhere.

pakrym-oai committed:
wip

pakrym-oai committed:
refactor: unify function/tool call error handling via FunctionCallError

- Handles errors in function/tool/cmd calls using `Result<String, FunctionCallError>` instead of wrapping outputs in `ResponseInputItem`.
- Removes unnecessary `ResponseInputItem` conversions and related From impls.
- Refactors exec handling, sandbox error, patch, plan/mcp tool call, and tool output mapping to propagate errors via Result.
- Demotes exec_command result_into_payload; now directly uses text output or error

pakrym-oai commented on codex-rs/core/src/function_tool.rs:
```diff
@@ -0,0 +1,7 @@
+use thiserror::Error;
+
+#[derive(Debug, Error, PartialEq)]
```
Plan to put more function-tool related code here.

pakrym-oai commented on codex-rs/core/src/apply_patch.rs:
```diff
@@ -17,7 +16,7 @@ pub(crate) enum InternalApplyPatchInvocation {
     /// of sandbox, because the user explicitly approved it. This is the
     /// result to use with the `shell` function call that contained `apply_patch`.
-    Output(ResponseInputItem),
+    Output(Result<String, FunctionCallError>),
```
This is the part that's I'm not sure is the cleanest.

pakrym-oai committed:
Merge branch 'main' into pakrym/cleanup-function-results

# Conflicts:
#	codex-rs/core/src/codex.rs

pakrym-oai committed:
codex: refactor MCP tool call handling to return structured output

Move conversion of MCP tool call results to structured output (ResponseInputItem) into handle_mcp_tool_call. Simplify error handling for invalid arguments and return errors as structured responses. Remove in-place stringification of results—conversion now occurs in higher layers.

jif-oai commented on codex-rs/core/src/codex.rs:
```diff
@@ -2370,40 +2410,25 @@ async fn handle_unified_exec_tool_call(
+    #[derive(Serialize)]
+    struct SerializedUnifiedExecResult<'a> {
+        session_id: Option<String>,
+        output: &'a str,
```
I love the usage of reference for sure. But I don't really understand why using it here?
In the end it goes in a serde_json::to_string which will have to copy the data and value is anyway dropped at the end of the function
This is of course totally fine but the lifetime specifier may just complexify the code for nothing

pakrym-oai commented on codex-rs/core/src/codex.rs:
```diff
@@ -2215,18 +2217,30 @@ async fn handle_response_item(
+                        content,
+                        success: Some(true),
+                    },
+                    Err(FunctionCallError::RespondToModel(msg)) => FunctionCallOutputPayload {
```
I don't want to make API types dependent on the agent loop types. Feels like the wrong dependency direction.

pakrym-oai committed:
core: own value of output in SerializedUnifiedExecResult

Changed SerializedUnifiedExecResult to own the output field (String) instead of borrowing (&str) to simplify serialization. Removed obsolete comments in related code.

jif-oai commented on codex-rs/core/src/codex.rs:
```diff
@@ -2414,15 +2439,10 @@ async fn handle_function_call(
     name: String,
     arguments: String,
     call_id: String,
```
Hmm it got shifted. Basically everywhere where you do `{e}`

jif-oai commented:
For the record: remaining points were discussed on Slack

pakrym-oai committed:
error: improve error formatting with debug output

- Use debug formatting ({err:?}, {output:?}, etc.) for error messages in function call handling and execution paths for more informative diagnostics.
